src/theory/arith/arith_msum.cpp

```diff
@@ -219,21 +219,26 @@ int ArithMSum::isolate(
   int ires = isolate(v, msum, veq_c, val, k);
   if (ires != 0)
   {
+    NodeManager* nm = NodeManager::currentNM();
     Node vc = v;
     if (!veq_c.isNull())
     {
       if (doCoeff)
       {
-        vc = NodeManager::currentNM()->mkNode(MULT, veq_c, vc);
+        vc = nm->mkNode(MULT, veq_c, vc);
       }
       else
       {
         return 0;
       }
     }
     bool inOrder = ires == 1;
-    veq = NodeManager::currentNM()->mkNode(
-        k, inOrder ? vc : val, inOrder ? val : vc);
+    // ensure type is correct for equality
+    if (k == EQUAL && !vc.getType().isInteger() && val.getType().isInteger())
```

Comment:
Is there a reason why it can't be the other way round? (vc is int but val is not)

Reply:
Yes, in this utility, `vc` is being solved for.  Thus, it would be *unsound* for an Int to be assigned a Real, which this utility sound ensure.  I will add an assertion for this.

Reply:
FYI, the assertion will be added on the PR that makes the type rules strict, it still doesn't hold after this PR quite yet.